---
title: "Exploring ways to analyze Anacapa biom tables"
author: "Gaurav Kandlikar"
date: "December 15, 2017"
output:
  html_document:
    code_folding: hide
---

```{r setup, include=FALSE}
library(knitr)
opts_chunk$set(echo = TRUE)
```

## File import and cleanup

The goal of this document is to begin analyses to benchmark Anacapa's performance to other eDNA/metagenomics analysis pipeline. We need to develop ways to compare the resulting site-frequency tables. 

First, let's import in some packages and a series of biom tables

```{r, message = F}
library(vegan); library(dplyr); library(stringr); library(knitr); library(tidyr); 
# biom_dir <- "CO1_biom_for_Gaurav_sanger/"
# Path to where the biom tables are stored
biom_dir <- "CO1_biom_crux/"
filenames <- list.files(path = biom_dir, pattern = ".biom")
# name of "true" biom table (or the one everything gets compared to)
mock_commuity_table <- "AA_CO1_mock_expected"

# read in all of the biom tables; save them in a list
biom_list <- lapply(as.list(filenames), function(x) 
  read.table(file = paste0(biom_dir,x), header = 1, sep = "\t", stringsAsFactors = F))
names(biom_list) <- str_replace(filenames, ".biom", "")
names(biom_list)[which(names(biom_list) == mock_commuity_table)] <- "mock_community"
```

## Verify structure of biom files
```{r}
structure_checker <- function(df) {
  if(typeof(df[,1] != "character")) {
    stop("First column of biom table should be the unique sequence name!")
  }
if(typeof(df[,ncol(df)] != "character")) {
    stop("Last column of biom table should be the taxonomy!")
  }

}

```

**Before going too far**, lets take a look at the structure of the files we imported. Here, I randomly sample two of the imported Biom tables to look at the structure of the file. I look at 2 just to get a sense of some of the variation out there.

```{r}
lapply(sample(biom_list, size = 2), tbl_df)


# May need to comment out this next chunk:
col_renamer <- function(df) {
  if(any(grepl("MiSeq", colnames(df)))) {
    return(rename(df, run1 = CO1_run1MiSeq300.16.S10.L001, run2 = CO1_run2MiSeq300.16.S10.L001))
  } else if (any(grepl("HiSeq", colnames(df)))) {
    return(rename(df, run1 = CO1_run1HiSeqTrim150.16.S10.L001, run2 = CO1_run2HiSeqTrim150.16.S10.L001))
  } else {
    return(df)
  }
}

biom_list <- lapply(biom_list,  col_renamer)

```


After having explored these files, it looks like there's (at least) a couple of things to fix... In some biom tables, we have `Taxonomic_ranks` that look like the following: `;;;;NA;`; `(blank)`. Let's reset the `NA`s to nothing; and let's replace the `(blank)` with `;;;;;`

```{r}
biom_list <- lapply(biom_list, function(x) x %>% tbl_df %>% 
                      mutate(sum.taxonomy = str_replace_all(sum.taxonomy, "NA", "")) %>%
                      mutate(sum.taxonomy = str_replace(sum.taxonomy, "\\(blank\\)", ";;;;;")) %>%
                      mutate(sum.taxonomy = str_replace(sum.taxonomy, "not_found", ";;;;;"))  %>% 
                      mutate(sum.taxonomy = str_replace(sum.taxonomy, "superkingdom:\\w+;", "")) %>% #  Gets rid of superkingdom
                      mutate(sum.taxonomy = str_replace(sum.taxonomy, "Eukaryota;", "")) %>% 
                      mutate(sum.taxonomy = str_replace(sum.taxonomy, "phylum:", "")) %>% 
                      mutate(sum.taxonomy = str_replace(sum.taxonomy, "class:", "")) %>% 
                      mutate(sum.taxonomy = str_replace(sum.taxonomy, "order:", "")) %>% 
                      mutate(sum.taxonomy = str_replace(sum.taxonomy, "family:", "")) %>% 
                      mutate(sum.taxonomy = str_replace(sum.taxonomy, "genus:", "")) %>% 
                      mutate(sum.taxonomy = str_replace(sum.taxonomy, "species:", "")))

```

Next, we filter out low-confidence reads (reads that are only showing up in very few places)
```{r}

# For absolute abundance, uncomment this chunk:
# threshold <- 100
# biom_list[["mock_community"]]$run1 = threshold+1
# biom_list[["mock_community"]]$run2 = threshold+1
  
# filter out the singletons
# biom_list <- lapply(biom_list, function(x) x %>% filter(run1 > threshold & run2 > threshold))

# filter out hits with low relative abundance
biom_list <- lapply(biom_list, function(x) x %>% mutate(rel_run1 = run1/sum(run1),
                                           rel_run2 = run2/sum(run2)))
rel_threshold <- .001
biom_list[["mock_community"]]$rel_run1 = rel_threshold*1.01
biom_list[["mock_community"]]$rel_run2 = rel_threshold*1.01

biom_list <- lapply(biom_list, function(x) x %>% filter(rel_run1 > rel_threshold & rel_run2 > rel_threshold))

```

### Summary stats on the biom tables

Let's run some numbers on what we are finding.

#### Number of taxa recognized per sample

The quickest thing to check is just the number of unique taxa found at each taxonomic rank (i.e. how many species did we ID? from how many genera?)

First, I separate out the `Taxonomic_path` column into six distinct columns (one each for phylum, class, order, family, genus, species):
```{r, warning=FALSE}
biom_list <- lapply(biom_list, function(x) 
  tidyr::separate(x, "sum.taxonomy" ,";", into = c("phylum", "class", "order", 
                                                   "family", "genus", "species")))

# Replace emptys with NAs

biom_list <- lapply(biom_list, function(x) x %>% mutate(phylum = ifelse(phylum == "", NA, phylum),
                                           class = ifelse(class == "", NA, class),
                                           order = ifelse(order == "", NA, order),
                                           family = ifelse(family == "", NA, family),
                                           genus = ifelse(genus == "", NA, genus),
                                           species = ifelse(species == "", NA, species)))
```

Then, we get the number of distinct groups in each column using the following dplyr code:
```{r}
N_taxa_per_biom <- t(sapply(biom_list, function(x) x %>% 
                              select(phylum:species) %>% summarize_all(n_distinct, na.rm = T)))
kable(N_taxa_per_biom, caption = "Number of distinct taxa reported in each biom table")
```


--------------

#### Taxa that are present in the True/Mock dataset but missing in biom:  
The first thing to look at will is a summary of which taxa in the Mock community are *missing* in the output biom tables, and which things we *are* finding in the biom tables that we don't expect to.

The following ugly function does all of that:

```{r}

find_missing_taxa <- function(biom, mock) {
  mock_sub <- mock %>% select(phylum:species)
  biom_sub <- biom %>% select(phylum:species)
  taxa_in_mock <- apply(mock_sub, 2, function(x) (unique(x))) 
  taxa_in_biom <- apply(biom_sub, 2, function(x) (unique(x))) 
  mock_but_not_sample <- sapply(1:length(taxa_in_biom), function(x)
    taxa_in_mock[[x]][!(taxa_in_mock[[x]] %in% taxa_in_biom[[x]])], simplify = T)
  sample_but_not_mock <- sapply(1:length(taxa_in_biom), function(x) 
    taxa_in_biom[[x]][!(taxa_in_biom[[x]] %in% taxa_in_mock[[x]])], simplify = T)
  num_mock_not_sample <- sapply(mock_but_not_sample, function(x) sum(!is.na(x)))
  num_sample_not_mock <- sapply(sample_but_not_mock, function(x) sum(!is.na(x)))

  names(mock_but_not_sample) <- colnames(biom_sub)
  names(sample_but_not_mock) <- colnames(biom_sub)
  names(num_mock_not_sample) <- colnames(biom_sub)
  names(num_sample_not_mock) <- colnames(biom_sub)

  return(list(mock_but_not_sample = mock_but_not_sample, sample_but_not_mock = sample_but_not_mock, 
              num_mock_not_sample = num_mock_not_sample, num_sample_not_mock = num_sample_not_mock))
}

comparisons <- lapply(biom_list, function(x) find_missing_taxa(x, biom_list[["mock_community"]]))
```

Now, the object `comparisons` is a list that has, per biom table, four pieces of information: `mock_but_not_sample`, `sample_but_not_mock`, `num_mock_not_sample`, `num_sample_not_mock`. 

A quick summary is just to see, numerically, how "off" each of the biom tables is to the mock community:

```{r, results='hold'}
kable(t(sapply(comparisons, '[[', 3)), 
             caption = "Number of taxa present in the Mock community 
             but not present in the analysed BIOM tables")

kable(t(sapply(comparisons, '[[', 4)), 
             caption = "Number of taxa present in the analysed BIOM tables 
             but not present in the Mock community")

```


Okay, so we seem to find all the phyla, class, and genera we expect to find, but sometimes miss an order or a family or a species. We always find a lot more taxa at each rank than were knowingly added to the mock. 

First, let's take a look at which taxa we do expect to find but don't find in each biom:
```{r}
mock_but_not_sample_print <- (t(data.frame(lapply(comparisons, function(x) 
  (sapply(1:6, function(y) paste(x$mock_but_not_sample[[y]] %>% na.omit %>% as.vector, collapse = "; ")))))))
colnames(mock_but_not_sample_print) <- c("phylum", "class", "order", "family", "genus", "species")
kable(mock_but_not_sample_print, caption = "Taxa present in Mock but missing in samples")
```

Clearly, it looks like we do a poor job of finding members of the family `Mysidae` in the order `Mysida`. It might be that the family is missing from CRUX, but the fact that we *do* find it in `CO1_trim_70` suggests it's present in CRUX... Interesting.

------------

Next, we can look at which taxa we *do* find but don't necessarily expect to. This output is harder to interpret... **Hover over the cells in the following table to view the contents**. 

```{r}
sample_but_not_mock_print <- (t(data.frame(lapply(comparisons, function(x) 
  (sapply(1:6, function(y) paste(x$sample_but_not_mock[[y]] %>% na.omit %>% as.vector, collapse = "; ")))))))
colnames(sample_but_not_mock_print) <- c("phylum", "class", "order", "family", "genus", "species")

```

<script>
$(document).ready(function(){
    $('[data-toggle="popover"]').popover(); 
});
</script>

```{r, message = F}
library(kableExtra)
popover_dt <- data.frame(
  position = names(biom_list),
  stringsAsFactors = FALSE
)
popover_dt$`phylum` <- cell_spec(
  paste("Phyla missing in Mock"), # Cell texts
  popover = spec_popover(
    content = sample_but_not_mock_print[,1],
    title = NULL,                           # title will add a Title Panel on top
    position = "auto"
  ))
popover_dt$`class` <- cell_spec(
  paste("Classes missing in Mock"), # Cell texts
  popover = spec_popover(
    content = sample_but_not_mock_print[,2],
    title = NULL,                           # title will add a Title Panel on top
    position = "auto"
  ))
popover_dt$`order` <- cell_spec(
  paste("Orders missing in Mock"), # Cell texts
  popover = spec_popover(
    content = sample_but_not_mock_print[,3],
    title = NULL,                           # title will add a Title Panel on top
    position = "auto"
  ))
popover_dt$`family` <- cell_spec(
  paste("Families missing in Mock"), # Cell texts
  popover = spec_popover(
    content = sample_but_not_mock_print[,4],
    title = NULL,                           # title will add a Title Panel on top
    position = "auto"
  ))
popover_dt$`genus` <- cell_spec(
  paste("Genera missing in Mock"), # Cell texts
  popover = spec_popover(
    content = sample_but_not_mock_print[,5],
    title = NULL,                           # title will add a Title Panel on top
    position = "auto"
  ))
popover_dt$`species` <- cell_spec(
  paste("Species missing in Mock"), # Cell texts
  popover = spec_popover(
    content = sample_but_not_mock_print[,6],
    title = NULL,                           # title will add a Title Panel on top
    position = "auto"
  ))
kable(popover_dt, "html", escape = FALSE) %>%
  kable_styling("striped", full_width = FALSE)

``` 

## Let's just print out the taxonomy...

```{r, comment="", results="asis"}
lapply(biom_list, function(x) x %>% arrange(phylum) %>% kable(., "html") %>%
  kable_styling() %>%
  scroll_box(height = "750px"))

```

